src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs

```diff
@@ -1252,5 +1245,236 @@ internal bool CanAddParsedValue(HttpHeaderParser parser)
+
+        // Used to store the CollectionsMarshal.GetValueRefOrAddDefault out parameter
+        private static bool s_dictionaryGetValueRefOrAddDefaultExistsDummy;
+
+        private const int InitialCapacity = 4;
+        internal const int ArrayThreshold = 64; // Above this threshold, header ordering will not be preserved
+
+        internal HeaderEntry[]? GetEntriesArray()
+        {
+            object? store = _headerStore;
+            if (store is null)
+            {
+                return null;
+            }
+            else if (store is HeaderEntry[] entries)
+            {
+                return entries;
+            }
+            else
+            {
+                return GetEntriesFromDictionary();
+            }
+
+            HeaderEntry[] GetEntriesFromDictionary()
+            {
+                var dictionary = (Dictionary<HeaderDescriptor, object>)_headerStore!;
+                var entries = new HeaderEntry[dictionary.Count];
+                int i = 0;
+                foreach (KeyValuePair<HeaderDescriptor, object> entry in dictionary)
+                {
+                    entries[i++] = new HeaderEntry
+                    {
+                        Key = entry.Key,
+                        Value = entry.Value
+                    };
+                }
+                return entries;
+            }
+        }
+
+        internal ReadOnlySpan<HeaderEntry> GetEntries()
+        {
+            return new ReadOnlySpan<HeaderEntry>(GetEntriesArray(), 0, _count);
+        }
+
+        internal int Count => _count;
+
+        private bool EntriesAreLiveView => _headerStore is HeaderEntry[];
+
+        private ref object GetValueRefOrNullRef(HeaderDescriptor key)
+        {
+            ref object valueRef = ref Unsafe.NullRef<object>();
+
+            object? store = _headerStore;
+            if (store is HeaderEntry[] entries)
+            {
+                for (int i = 0; i < _count && i < entries.Length; i++)
+                {
+                    if (key.Equals(entries[i].Key))
+                    {
+                        valueRef = ref entries[i].Value;
+                        break;
+                    }
+                }
+            }
+            else if (store is not null)
+            {
+                valueRef = ref CollectionsMarshal.GetValueRefOrNullRef(Unsafe.As<Dictionary<HeaderDescriptor, object>>(store), key);
+            }
+
+            return ref valueRef;
+        }
+
+        private ref object? GetValueRefOrAddDefault(HeaderDescriptor key)
+        {
+            object? store = _headerStore;
+            if (store is HeaderEntry[] entries)
+            {
+                for (int i = 0; i < _count && i < entries.Length; i++)
+                {
+                    if (key.Equals(entries[i].Key))
+                    {
+                        return ref entries[i].Value!;
+                    }
+                }
```

Comment:
Do we need to open-code these comparison loops, or could we use IndexOf?

Reply:
Since we're searching against `entries[i].Key` and not the whole type, we would need a custom `IndexOf` method anyway.
Extracting that logic and replacing the 3 loops here with it, the performance regresses by ~5% while increasing LOC.

Reply:
> we would need a custom IndexOf method anyway

Just a custom comparer, no?

If it's not a good tradeoff, fine. I just want us to strongly prefer using built-in functionality whenever possible.